include/utilities/bmi_utilities.hpp

```diff
@@ -0,0 +1,110 @@
+#ifndef NGEN_BMI_UTILITIES_HPP
+#define NGEN_BMI_UTILITIES_HPP
+
+#include <string>
+#include <vector>
+#include <boost/type_index.hpp>
+#include "Bmi_Adapter.hpp"
+
+namespace models {
+    namespace bmi {
+        namespace helper {
+
+            /**
+             * @brief make a vector of type @tparam TO of size @param count. Copy elements of @tparam FROM type in data into the vector.
+             * 
+             * This function relies on the std library to cast data approriately during the construction of the vector.
+             * 
+             * @tparam TO data type to cast elements of the data pointer to
+             * @tparam FROM data type of the pointer data is extracted from
+             * @param data pointer to @param count number of @tparam FROM type elements
+             * @param count number of elements in the @param data pointer
+             * @return std::vector<TO> vector filled with @param count number of elements extracted from @param data
+             */
+            template<typename TO, typename FROM>
+            static inline std::vector<TO> make_vector(const FROM* data, const size_t& count){
+                //Let Return Value Optimization (move semantics) help here (no copy of vector)
+                return std::vector<TO>(data, data+count);
+            }
+        }
+
+        /**
+         * @brief Copy values from a BMI model adapter and box them into a vector.
+         * 
+         * @tparam T Type to cast data from the BMI model to
+         * @tparam A Bmi model type
+         * @param model Bmi model adapter that interfaces to @tparam A bmi
+         * @param name Bmi variable name to query the model for
+         * @return std::vector<T> Copy of data from the BMI model for variable @param name
+         */
+        template <typename T, typename A>
+        std::vector<T> GetValue(Bmi_Adapter<A>& model, const std::string& name) {
+            //TODO make model const ref
+            int total_mem = model.GetVarNbytes(name);
+            int item_size = model.GetVarItemsize(name);
+            int num_items = total_mem/item_size;
+            //Determine what type we need to cast from
+            std::string type = model.get_analogous_cxx_type(model.GetVarType(name), item_size);
+            
+            //C++ form of malloc
+            void* data = ::operator new(total_mem);
```

Comment:
maybe attach smart pointer to this to make sure that memory is not leaked in the case of an exception. We have in general avoided new/delete malloc/free as much as possible.

Reply:
This is partly due to getting a smart pointer of the correct size which also properly deletes the correct amount of memory.  A `void*` of the correct size is required by BMI to populate the array in `GetValue`.  I recall trying to do this with a smart pointer and having issues with it, but I can't remember off the top of my head exactly what the complication was.  It very well may have been getting it typed as `void*` with the correct number of bytes allocated.

The use of manual allocation/cleanup here is at least isolated to this function, and we can fix the potential leak in the exception throwing.  If we can do this sufficiently with a smart pointer, we can change this.

Reply:
Interesting question. Maybe you ran into a problem using `unique_ptr<void>`? Maybe I'll play with this.

Reply:
This fixes the memory link anytime there was an error, which was my only concern with the original request.